File: src/ovi_voice_assistant/stt/stt.py
"""Speech-to-text base class."""

from abc import ABC, abstractmethod
from collections.abc import AsyncIterator, Awaitable, Callable

from ovi_voice_assistant.config import Settings

# Callback fired when VAD detects speech start
VadStartCallback = Callable[[], Awaitable[None]]


class STT(ABC):
    @abstractmethod
    def __init__(self, settings: Settings) -> None: ...

    @abstractmethod
    def load(self) -> None: ...

    @abstractmethod
    def transcribe(self, audio_bytes: bytes) -> str: ...

    @abstractmethod
    async def transcribe_stream(
        self,
        audio_chunks: AsyncIterator[bytes],
        on_vad_start: VadStartCallback | None = None,
    ) -> str: ...

File: internal/stt/stt.go
// Package stt implements speech-to-text providers with Silero VAD
// end-of-speech detection.
package stt

import (
	"context"
	"fmt"

	"github.com/bryfur/ovi-voice-assistant/internal/config"
)

// VADStartCallback fires when VAD detects speech start.
type VADStartCallback func()

// STT is a speech-to-text engine.
type STT interface {
	// Load loads models; must be called before transcription.
	Load() error
	// Transcribe transcribes a complete 16 kHz mono PCM buffer.
	Transcribe(pcm []byte) (string, error)
	// TranscribeStream consumes mic audio until end-of-speech and returns
	// the transcript. onVADStart may be nil.
	TranscribeStream(ctx context.Context, chunks <-chan []byte, onVADStart VADStartCallback) (string, error)
}

// Create builds the configured STT provider.
func Create(settings *config.Settings) (STT, error) {
	switch settings.STT.Provider {
	case "whisper":
		return NewWhisperSTT(settings), nil
	case "nemotron":
		return NewNemotronSTT(settings), nil
	}
	return nil, fmt.Errorf("unknown STT provider: %s", settings.STT.Provider)
}
